File: algorithm/graph/UnDirectedGraph.java
/**
 * UnDirectedGraph
 */

package algorithm.graph;

public class UnDirectedGraph extends AbstGraph {
    private int size;
    private Node[] nodes;

    public UnDirectedGraph(int size) {
        this.size = size;
        this.nodes = new Node[size];

        for (int i=0; i<size; i++) {
            nodes[i] = new Node(i);
        }
    }

    /**
     * addEdge
     *  - 두 노드간 간선 추가
     */
    @Override
    public void addEdge(int a, int b) {
        if ((a < 0) || (b < 0) || 
            (this.size < a) || (this.size < b)) {

            throw new Error();
        }

        Node n1 = nodes[a];
        Node n2 = nodes[b];

        if (!n1.adjacent.contains(n2)) {
            n1.adjacent.add(n2);
        }
        if (!n2.adjacent.contains(n1)) {
            n2.adjacent.add(n1);
        }
    }

    public int getSize() {
        return this.size;
    }

    public Node getNode(int idx) {
        return this.nodes[idx];
    }
}

File: python/algorithm/graph/undirected_graph.py
from .graph_base import GraphBase


Node = GraphBase.Node


class UnDirectedGraph(GraphBase):
    def __init__(self, size: int):
        self._size = size
        self._nodes = []

        for i in range(self._size):
            self._nodes.append(Node(i))

    def add_edge(self, from_: int, to_: int):
        if (from_ < 0) or (to_ < 0) or \
                (self._size < from_) or (self._size < to_):
            raise IndexError("")

        n1 = self.node_at(from_)
        n2 = self.node_at(to_)

        if n2 not in n1.adjacent:
            n1.adjacent.append(n2)
        if n1 not in n2.adjacent:
            n2.adjacent.append(n1)

    def node_at(self, idx: int) -> Node:
        return self._nodes[idx]

    @property
    def size(self) -> int:
        return self._size
